misk/src/main/kotlin/misk/web/PemComboFile.kt

```diff
@@ -0,0 +1,144 @@
+package misk.web
+
+import okio.Buffer
+import okio.BufferedSource
+import okio.ByteString
+import okio.Okio
+import java.io.File
+import java.io.IOException
+import java.security.KeyFactory
+import java.security.KeyStore
+import java.security.cert.Certificate
+import java.security.cert.CertificateFactory
+import java.security.spec.KeySpec
+import java.security.spec.PKCS8EncodedKeySpec
+import java.security.spec.RSAPrivateCrtKeySpec
+import java.security.GeneralSecurityException
+import sun.security.util.DerInputStream
+
+/**
+ * A file containing a mix of PEM-encoded certificates and PEM-encoded private
+ * keys. Can be used both for trust stores (which certificate authorities a TLS
+ * client trusts) and also for TLS servers (which certificate chain a TLS server
+ * serves).
+ */
+data class PemComboFile(
+  private val certificates: List<ByteString>,
+  private val privateRsaKeys: List<ByteString>,
+  private val privateKeys: List<ByteString>,
+  private val passphrase: String
+) {
+  fun tlsClientTrustStore(): KeyStore {
```

Comment:
Not sure I understand why we have these two functions - why not just one that parses whatever is in the file and builds a keystore out of the contents?

Reply:
Specifically, the distinction between a "trust store" and a "key store" is an artificial one; it's just how the application code wants to use the contents of the file. It's a distinction that matters at the web server / client level, but shouldn't be something that permeates a basic file parser.

Reply:
I recommended these, but you’re right that they’re wrong!

I still think we need something to help callers to differentiate our two main use cases:

 - 1 private key & _N_ certs that I use to prove my own identity
 - 0 private keys and _N_ certs that I trust when validating other services’ identity

One thing that makes this extra tricky is that we use the private key both for clients and servers. Typically with TLS it’s just the private key for servers.

Reply:
I'd propose doing this with higher level APIs that wrap an underlying ``KeyStore`` and validate that they contain / expose only the expected contents e.g. have a ``CertStore`` that exposes the advertised certificate and private key and a ``TrustStore`` that exposes only a collection of trusted certificate chains. That way the PEM parser just builds a general ``KeyStore`` out of its contents (whatever they may be) and then the SSL layers enforce particular constraints on the contents of a ``KeyStore`` depending on the use case.

Reply:
I like that concept, I'll add some higher level classes that serve as very basic wrappers for these two store types. We can iterate them into more general abstractions down the road.